scripts/run_all.py

```diff
@@ -4,86 +4,239 @@
 
 import argparse
 import os
+import socket
 import subprocess
 import time
 from glob import glob
-from typing import Any, List
+from types import TracebackType
+from typing import Any, List, Optional, Tuple, Type, cast
 
 
-def run_py_example(path: str, args: List[str] = []) -> None:
+def run_py_example(path: str, viewer_port: int, wait: bool = True, save: Optional[str] = None) -> Any:
+    args = ["--connect", "--addr", f"127.0.0.1:{viewer_port}"]
+    if save is not None:
+        args += ["--save", str(save)]
+
     process = subprocess.Popen(
         ["python3", "main.py", "--num-frames=30", "--steps=200"] + args,
         cwd=path,
     )
-    returncode = process.wait()
-    print(f"process exited with error code {returncode}")
+    if wait:
+        returncode = process.wait()
+        print(f"process exited with error code {returncode}")
```

Comment:
Any non-zero returncode should result in `run_all.py` returning non-zero, and printing a big fat error message

Reply:
Changed this to assert instead. Not sure how I should do the same when running `--separate`.